pkg/api/validate/dynamic/sku.go

```diff
@@ -0,0 +1,71 @@
+package dynamic
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/computeskus"
+)
+
+// ValidateWorkerSku uses resourceSkusClient to ensure that the VM sizes listed in the cluster document are available for use in the target region.
+func (dv *dynamic) ValidateVmSku(ctx context.Context, location string, subscriptionID string, oc *api.OpenShiftCluster) error {
+	// Get a list of available worker SKUs, filtering by location.
+	filter := fmt.Sprintf("location eq %s", location)
+	skus, err := dv.resourceSkusClient.List(ctx, filter)
+	if err != nil {
+		return err
+	}
+
+	filteredSkus := computeskus.FilterVmSizes(skus, location)
+
+	// Ensure sku in MasterProfile exists in target region
+	masterSkuAvailable := false
+	masterProfileSku := string(oc.Properties.MasterProfile.VMSize)
+
+	if filteredSkus[masterProfileSku] == nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is unavailable in region '%v'", masterProfileSku, location)
+	} else {
+		masterSkuAvailable = true
+	}
+
+	// Fail if sku is available, but restricted within the subscription. Restrictions are subscription-specific.
+	// https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/error-sku-not-available
+	for _, restriction := range *filteredSkus[masterProfileSku].Restrictions {
+		restrictedLocation := strings.Join(*restriction.RestrictionInfo.Locations, "")
+
+		if restrictedLocation == location &&
+			masterSkuAvailable {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.masterProfile.VMSize", "The selected SKU '%v' is restricted in region '%v' for selected subscription", masterProfileSku, location)
+		}
+	}
+
+	// In case there are multiple WorkerProfiles listed in the cluster document (such as post-install),
+	// compare VMSize in each WorkerProfile to the resourceSkusClient call above to ensure that the sku is available in region.
+	for _, workerprofile := range oc.Properties.WorkerProfiles {
+		workerSkuAvailable := false
+		workerProfileSku := string(workerprofile.VMSize)
+
+		if filteredSkus[workerProfileSku] == nil {
+			return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "properties.workerProfile.VMSize", "The selected SKU '%v' is unavailable in region '%v'", workerProfileSku, location)
```

Comment:
Must be `properties.workerProfiles[%d].blahBlah` because `workerProfiles` is a slice. Same applies to the error below.